File: erlang/src/p047.erl
-module(p047).

-export([start/0]).

f() ->
  SPrimes = sets:from_list(utils:primes(1000)),
  g(1, 2, 3, 4, SPrimes).

g(A, B, C, D, SPrimes) ->
  NoMatch = lists:any(fun(X) -> h(X, SPrimes) /= 4 end, [A, B, C, D]),
  if
    NoMatch -> g(A + 1, B + 1, C + 1, D + 1, SPrimes);
    true -> A
  end.

h(N, SPrimes) ->
  Divisors = sets:from_list(utils:divisors(N)),
  PrimeDivisors = sets:to_list(sets:intersection(Divisors, SPrimes)),
  length(PrimeDivisors).

start() ->
  io:format("~p ~n", [f()]).

File: go/src/p047/p047.go
package main

import (
    "fmt"
    "utils"
)

func main() {
    primes := utils.Primes(1000)

    for i := uint64(2); ; i++ {
        if len(utils.DistinctFactors(i, primes)) == 4 &&
           len(utils.DistinctFactors(i + 1, primes)) == 4 &&
           len(utils.DistinctFactors(i + 2, primes)) == 4 &&
           len(utils.DistinctFactors(i + 3, primes)) == 4 {
            fmt.Println(i)
            break
        }
    }
}
